p2p/host/basic/basic_host.go

```diff
@@ -377,9 +390,43 @@ func (h *BasicHost) Connect(ctx context.Context, pi pstore.PeerInfo) error {
 		return nil
 	}
 
+	resolved, err := h.resolveAddrs(ctx, h.Peerstore().PeerInfo(pi.ID))
+	if err != nil {
+		return err
+	}
+	h.Peerstore().AddAddrs(pi.ID, resolved, pstore.TempAddrTTL)
+
 	return h.dialPeer(ctx, pi.ID)
 }
 
+func (h *BasicHost) resolveAddrs(ctx context.Context, pi pstore.PeerInfo) ([]ma.Multiaddr, error) {
+	proto := ma.ProtocolWithCode(ma.P_IPFS).Name
+	p2paddr, err := ma.NewMultiaddr("/" + proto + "/" + pi.ID.Pretty())
+	if err != nil {
+		return nil, err
+	}
+
+	var addrs []ma.Multiaddr
+	for _, addr := range pi.Addrs {
+		addrs = append(addrs, addr)
+
+		reqaddr := addr.Encapsulate(p2paddr)
+		resaddrs, err := h.maResolver.Resolve(ctx, reqaddr)
```

Comment:
I think it would better if there was an address match check before blanket attempting to resolve.

Also, it's not clear what is the behaviour of `Resolve` on non dns address. 
For example, will it resolve to an empty address list if it is not a dns address? will it error out? or will it repeat its input?